### PR TITLE
feat: eip-712 structured signing

### PR DESCRIPTION
**Pull request by gomesalexandre**
- [x] KK (already implemented and confirmed working)
- [x] Native (already implemented)
- [x] MM multichain
- [x] Coinbase
- [x] Ledger 
- [x] Native (already implemented)
- [x] not authored for MM, this adapter is not used anymore, we should probably remove it, and while at it renames its `@shapeshiftoss/hdwallet-shapeshift-multichain` name to `@shapeshiftoss/hdwallet-metamask-shapeshift-multichain` if that's possible, seems like an oversight when this package was authored~~MM~~

### Testing

Test me with https://github.com/shapeshift/web/pull/7420

- Ensure MM and Coinbase Wallet now display typed data when signing CoW orders signing
- Ensure native is still happy when signing CoW orders
- Ensure KK is still happy with CoW, and now displays *slightly* more data than before (i.e hash domain and order, vs. only a hashed message signing on develop)
- Ensure Ledger is happy with CoW, and now displays *slightly* more data than before (i.e domain version and chainId vs. only a hashed message signing on device on develop)

### Native

image

https://explorer.cow.fi/arb1/orders/0x11bd0702c9119f88985a468a2fea636e87f3963f185f60701d06c3ddbfd9f17f5daf465a9ccf64deb146eeae9e7bd40d6761c98666a24041?tab=overview

### MetaMask

image
image

https://explorer.cow.fi/arb1/orders/0xe5480460293ab537525267377dc397157c1a0a24033c2007b19f350f3fc0ce0d5daf465a9ccf64deb146eeae9e7bd40d6761c98666a2411a?tab=overview

### Ledger

image

https://explorer.cow.fi/orders/0x8fd2cdb01900ec4392d269224b7f0891e8cae4a6dd1ca8c1fccf82e0a92615805daf465a9ccf64deb146eeae9e7bd40d6761c98666a2419f?tab=overview

### KeepKey

image

https://explorer.cow.fi/orders/0x56c858725816f2511aafd07ded2dc890efa9d250e37eb3038b79d7b6c4aa54d85daf465a9ccf64deb146eeae9e7bd40d6761c98666a2422b?tab=overview

### Coinbase

image
image

https://explorer.cow.fi/arb1/orders/0x6997005bc0b185b06e8617af6acadf16d581f7e46c4561f797ec8c934f53591a5daf465a9ccf64deb146eeae9e7bd40d6761c98666a2429c?tab=overview

### CODE REVIEW

File: examples/sandbox/package.json
Comment (gomesalexandre) [inline examples/sandbox/package.json:1]: Obviously remove and revert all the verdaccio-y bits and run proper lerna script before authoring

File: packages/hdwallet-coinbase/src/ethereum.ts [hunk 1]
```diff
@@ -80,6 +80,34 @@ export async function ethVerifyMessage(msg: core.ETHVerifyMessage, ethereum: any
   return null;
 }
 
+export async function ethSignTypedData(
+  msg: core.ETHSignTypedData,
+  ethereum: any,
+  address: string
+): Promise<core.ETHSignedMessage | null> {
+  try {
+    // eslint-disable-next-line no-console
```
Comment (gomesalexandre) [inline packages/hdwallet-coinbase/src/ethereum.ts:89]: revert before opening me

File: packages/hdwallet-metamask-shapeshift-multichain/src/ethereum.ts [hunk 2]
```diff
@@ -87,6 +87,34 @@ export async function ethSignMessage(
   }
 }
 
+export async function ethSignTypedData(
+  msg: core.ETHSignTypedData,
+  ethereum: any,
+  address: string
+): Promise<core.ETHSignedMessage | null> {
+  try {
+    // eslint-disable-next-line no-console
```
Comment (gomesalexandre) [inline packages/hdwallet-metamask-shapeshift-multichain/src/ethereum.ts:96]: here too

File: packages/hdwallet-metamask-shapeshift-multichain/src/shapeshift-multichain.ts [hunk 3]
```diff
@@ -784,8 +780,13 @@ export class MetaMaskShapeShiftMultiChainHDWallet
   }
 
   public async ethSignMessage(msg: core.ETHSignMessage): Promise<core.ETHSignedMessage | null> {
-    const signedMessage = await this.ethGetAddress(this.provider);
-    return signedMessage ? Eth.ethSignMessage(msg, this.provider, signedMessage) : null;
+    const address = await this.ethGetAddress(this.provider);
+    return address ? Eth.ethSignMessage(msg, this.provider, address) : null;
+  }
```
Comment (gomesalexandre) [inline packages/hdwallet-metamask-shapeshift-multichain/src/shapeshift-multichain.ts:785]: diff is drunk, this isn't diffed, but moved under, no idea why this was living in this random location above

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**